File: packages/test-framework/config/ConfigManager.ts
import * as fs from 'fs';
import * as path from 'path';

export interface EnvConfig {
  environment: string;
  baseUrl: string;
  apiBaseUrl: string;
  credentials: Record<string, string>;
  variables: Record<string, any>;
}

export class ConfigManager {
  private static instance: EnvConfig | null = null;

  /**
   * Loads and parses the environment configuration. Resolves dynamic placeholders from environment variables.
   */
  public static getConfig(): EnvConfig {
    if (this.instance) {
      return this.instance;
    }

    const env = process.env.ENV || 'qa';
    const configPath = path.join(process.cwd(), 'resources', 'config', 'environments', `${env}.json`);

    if (!fs.existsSync(configPath)) {
      throw new Error(`Configuration file not found for environment "${env}" at: ${configPath}`);
    }

    const rawContent = fs.readFileSync(configPath, 'utf8');
    let configObj: EnvConfig;
    try {
      configObj = JSON.parse(rawContent);
    } catch (err: any) {
      throw new Error(`Failed to parse configuration file at ${configPath}: ${err.message}`);
    }

    // Resolve system environment variables dynamically: e.g., ${QA_USERNAME}
    const resolvedConfig = this.resolveEnvVars(configObj) as EnvConfig;
    this.instance = resolvedConfig;
    return resolvedConfig;
  }

  private static resolveEnvVars(obj: any): any {
    if (typeof obj === 'string') {
      return obj.replace(/\${(\w+)}/g, (_, varName) => {
        return process.env[varName] !== undefined ? process.env[varName]! : `\${${varName}}`;
      });
    } else if (Array.isArray(obj)) {
      return obj.map(item => this.resolveEnvVars(item));
    } else if (obj !== null && typeof obj === 'object') {
      const result: Record<string, any> = {};
      for (const key in obj) {
        result[key] = this.resolveEnvVars(obj[key]);
      }
      return result;
    }
    return obj;
  }
}

export const config = ConfigManager.getConfig();

File: tests/support/config_manager.py
import json
import os
import re
from pathlib import Path
from typing import Any


class ConfigManager:
  """Load WebPilot environment JSON (resources/config/environments/<env>.json)."""

  _instance: dict[str, Any] | None = None

  @classmethod
  def get_config(cls) -> dict[str, Any]:
    if cls._instance is not None:
      return cls._instance
    env = os.environ.get("ENV", "qa")
    config_path = Path.cwd() / "resources" / "config" / "environments" / f"{env}.json"
    if not config_path.exists():
      raise FileNotFoundError(f'Configuration file not found for environment "{env}": {config_path}')
    raw = json.loads(config_path.read_text(encoding="utf-8"))
    cls._instance = cls._resolve_env_vars(raw)
    return cls._instance

  @classmethod
  def _resolve_env_vars(cls, obj: Any) -> Any:
    if isinstance(obj, str):
      return re.sub(
        r"\$\{(\w+)\}",
        lambda match: os.environ.get(match.group(1), match.group(0)),
        obj,
      )
    if isinstance(obj, list):
      return [cls._resolve_env_vars(item) for item in obj]
    if isinstance(obj, dict):
      return {key: cls._resolve_env_vars(value) for key, value in obj.items()}
    return obj


config = ConfigManager.get_config()
